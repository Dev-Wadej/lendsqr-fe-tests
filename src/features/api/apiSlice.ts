import {createApi,fetchBaseQuery} from '@reduxjs/toolkit/query/react';

interface IUsersData{
    createdAt:string;
    orgName:string
   email:string;
   id:string;
   userName:string;
   phoneNumber:string;
  }
interface IEachUser{
    accountNumber: string;
    email:string;
    profile: {
        firstName: string;
        lastName: string;
        phoneNumber: string;
        avatar:string;
    gender: string;
    bvn: string;
    }
    guarantor: {
        firstName: string,
        lastName: string,
        phoneNumber: string;
        
      },
      education: {
        level: string;
        employmentStatus: string;
        sector: string;
        duration: string;
        officeEmail: string;
        monthlyIncome: string[];
        loanRepayment: string;

      },
      socials: {
        facebook: string;
        instagram: string;
        twitter: string;
      },

}

export const apiSlice = createApi({
    reducerPath: 'api',
    baseQuery: fetchBaseQuery({ baseUrl: 'https://6270020422c706a0ae70b72c.mockapi.io/lendsqr/api/v1' }),
    tagTypes: ['Users'],
    endpoints: (builder) => ({
        getUsers: builder.query<IUsersData[],boolean>({
            query: () => '/users',
            providesTags: ['Users'],
        }),
        getEachUser: builder.query<IEachUser,string>({
            query: (id) => `/users/${id}`,
            providesTags: ['Users'],
        }),
    }),
    
})

export const {
    useGetUsersQuery,useGetEachUserQuery
} = apiSlice