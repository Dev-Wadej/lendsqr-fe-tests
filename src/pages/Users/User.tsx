import {useState} from 'react'
//============== Custom hooks======================================
import {useGetUsersQuery} from '../../features/api/apiSlice'
import {useSidebar } from '../../context/ContextProvider'
import { useUsersPagination } from '../../hooks/useUsersPagination'
//=================================================================

//========================= SVG======================================
import {ReactComponent as Users} from '../../data/svgs/users.svg'
import {ReactComponent as ActiveUsers} from '../../data/svgs/activeusers.svg'
import {ReactComponent as UsersWithLoan} from '../../data/svgs/userswithloan.svg'
import {ReactComponent as UsersWithSavings} from '../../data/svgs/userswithsavings.svg'
import {ReactComponent as DropDown} from '../../data/svgs/dropdown.svg'
//=================================================================

//========================= Component======================================
import EachUser from '../../components/EachUser/EachUser'
import FormDropDown from '../../components/FormDropDown/FormDropDown'
import MenuIcon from '@mui/icons-material/Menu';
import PaginationRounded from '../../components/Pagination/Pagination'
//======================================================================

import './user.scss'

const User = () => {

  const dashboardFilterActions= ['ORGANIZATION','USERNAME','EMAIL','PHONE NUMBER','DATE JOINED','STATUS']
  const boxesProperties = [
    {name:'USERS', icon:Users, amount:'2,453'},
    {name:'ACTIVE USERS',icon:ActiveUsers, amount:'2,453'},
    {name:'USERS WITH LOAN', icon:UsersWithLoan, amount:'12,453'},
    {name:'USERS WITH SAVINGS', icon:UsersWithSavings, amount:'102,453'},
  ]

  const {setSidebar,sidebar,pageNo}=useSidebar()
  const [showFilterForm,setShowFilterForm]=useState(false)


  

  const { data, isLoading,isError } = useGetUsersQuery(true)
  const res=useUsersPagination?.(data!,pageNo)  
    
  const handleOpenSideBar=()=>{
    setSidebar?.(true)
  }
 
  const handleToggleFilterForm=()=>{
    setShowFilterForm((prev)=>!prev)
  }

  
  if(isLoading){
    return(
      <div>Loading</div>
    )
  }
  if(isError){
    return(
      <div>Error Messgae</div>
    )
  }

  return (
    <section className='all-users'>
      <div onClick={handleOpenSideBar}>
      { !sidebar ? <MenuIcon className='menu'/> : <></>}
      </div>
      <h2>Users</h2>
      <ul className='users-box'>
          {boxesProperties.map(dataBoxesProp=>(
            <li>
            <span className='user-box-icon'> <dataBoxesProp.icon/></span>
            <span className='users-box-title'>{dataBoxesProp.name}</span>
            <span className='users-box-amount'>{dataBoxesProp.amount}</span>
            </li>

          ))}
      </ul>
      <div className='main-dashboard-wrapper'>

      <ul className='main-dashboard'>
        {showFilterForm &&
        <FormDropDown/>
        }
        {dashboardFilterActions.map(callToAction=>(
          <li onClick={handleToggleFilterForm}>
          <span>{callToAction}</span>
          <span><DropDown/></span>
          </li>
        ))}
      </ul>
      <div className='status-down-wrapper'>
      {res?.amtPerPage?.map((eachUserDetails,idx)=>(
          <EachUser key={idx} data={eachUserDetails} />
      ))}
 </div>
      </div>
      <PaginationRounded totalPage={res?.totalPages}/>

    </section>
  )
}

export default User
