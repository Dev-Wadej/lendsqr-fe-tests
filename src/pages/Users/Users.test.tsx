import User from './User'
import renderer from 'react-test-renderer';
import {cleanup,} from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { ApiProvider } from '@reduxjs/toolkit/dist/query/react';
import { apiSlice } from '../../features/api/apiSlice';



afterEach(cleanup)
test('User component snapshot HTML', () => {
    const output = renderer.create(<MemoryRouter>
        <ApiProvider api={apiSlice}>
       <User/>
        </ApiProvider>
    </MemoryRouter>  ).toJSON();
  expect(output).toMatchSnapshot();
});