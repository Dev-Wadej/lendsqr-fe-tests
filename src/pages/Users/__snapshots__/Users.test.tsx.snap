// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User component snapshot HTML 1`] = `
<div>
  Loading
</div>
`;
